cpp/src/quantiles/groupby.cu

```diff
@@ -0,0 +1,264 @@
+                          _key_table.num_columns(), _key_sorted_order.get(),
+                          &context));
+  } else {  // Pandas style
+    auto key_cols_bitmask = row_bitmask(_key_table);
+
+    gdf_column modified_fist_key_col = *(_key_table.get_column(0));
+    modified_fist_key_col.valid =
+        reinterpret_cast<gdf_valid_type*>(key_cols_bitmask.data().get());
+
+    auto keys = const_cast<cudf::table*>(&_key_table);
+    std::vector<gdf_column*> modified_key_cols_vect(keys->begin(), keys->end());
+    modified_key_cols_vect[0] = &modified_fist_key_col;
+    cudf::table modified_key_col_table(modified_key_cols_vect.data(),
+                                      modified_key_cols_vect.size());
+
+    gdf_context temp_ctx;
+    temp_ctx.flag_null_sort_behavior = GDF_NULL_AS_LARGEST;
+
+    CUDF_TRY(gdf_order_by(modified_key_col_table.begin(), nullptr,
+                          modified_key_col_table.num_columns(),
+                          _key_sorted_order.get(), &temp_ctx));
+
+    CUDF_TRY(gdf_count_nonzero_mask(
+        reinterpret_cast<gdf_valid_type*>(key_cols_bitmask.data().get()),
+        _key_table.num_rows(),
+        &_num_keys));
+  }
+
+  return *_key_sorted_order;
+}
+
+rmm::device_vector<gdf_size_type> const& groupby::group_indices() {
+  if (_group_ids)
+    return *_group_ids;
+
+  index_vector idx_data(_num_keys);
+
+  auto counting_iter = thrust::make_counting_iterator<gdf_size_type>(0);
+  auto device_input_table = device_table::create(_key_table);
+  bool nullable = device_input_table.get()->has_nulls();
+  auto sorted_order = reinterpret_cast<gdf_size_type*>(key_sort_order().data);
+  decltype(idx_data.begin()) result_end;
+
+  if (nullable) {
+    auto comp = row_equality_comparator<true>(*device_input_table, true);
+    result_end = thrust::unique_copy(
+      thrust::device, counting_iter, counting_iter + _num_keys,
+      idx_data.begin(), transform_row_eq_comparator<true>{comp, sorted_order});
+  } else {
+    auto comp = row_equality_comparator<false>(*device_input_table, true);
+    result_end = thrust::unique_copy(
+      thrust::device, counting_iter, counting_iter + _num_keys,
+      idx_data.begin(), transform_row_eq_comparator<false>{comp, sorted_order});
+  }
+
+  gdf_size_type num_groups = thrust::distance(idx_data.begin(), result_end);
+  _group_ids = std::make_unique<index_vector>(idx_data.begin(), idx_data.begin() + num_groups);
+
+  return *_group_ids;
+}
+
+rmm::device_vector<gdf_size_type> const& groupby::group_labels() {
+  if (_group_labels)
+    return *_group_labels;
+
+  // Get group labels for future use in segmented sorting
+  _group_labels = std::make_unique<index_vector>(_num_keys);
+
+  auto& group_labels = *_group_labels;
+  thrust::fill(group_labels.begin(), group_labels.end(), 0);
+  auto group_labels_ptr = group_labels.data().get();
+  auto group_ids_ptr = group_indices().data().get();
+  thrust::for_each_n(thrust::make_counting_iterator(1),
```

Comment:
RMM exec policy (even though this probably doesn't allocate, better to be consistent).